Source/AwsCommonRuntimeKit/auth/credentials/CRTAWSCredentialsProvider.swift

```diff
@@ -226,7 +231,39 @@ public final class CRTAWSCredentialsProvider {
         return future
     }
 
+    static func setUpShutDownOptions(shutDownOptions: CRTCredentialsProviderShutdownOptions?)
+    -> aws_credentials_provider_shutdown_options {
+
+        let pointer: UnsafeMutablePointer<CRTCredentialsProviderShutdownOptions>? = fromOptionalPointer(ptr: shutDownOptions)
+        let shutDownOptionsC = aws_credentials_provider_shutdown_options(shutdown_callback: { userData in
+            guard let userData = userData else {
+                return
+            }
+            let pointer = userData.assumingMemoryBound(to: CRTCredentialsProviderShutdownOptions.self)
+            defer {pointer.deinitializeAndDeallocate()}
+            pointer.pointee.shutDownCallback()
+
+        }, shutdown_user_data: pointer)
+
+        return shutDownOptionsC
+    }
+
     deinit {
         aws_credentials_provider_release(rawValue)
     }
 }
+
+private func getCredentialsDelegateFn(_ delegatePtr: UnsafeMutableRawPointer?,
+                                      _ callbackFn: (@convention(c)(OpaquePointer?, Int32, UnsafeMutableRawPointer?) -> Void)?,
+                                      _ userData: UnsafeMutableRawPointer?) -> Int32 {
+    guard let credentialsProvider = delegatePtr?.assumingMemoryBound(to: CRTCredentialsProvider.self) else {
+        return 1
+    }
+    guard let credentialCallbackData = userData?.assumingMemoryBound(to: CRTCredentialsProviderCallbackData.self) else {
+        return 1
+    }
+
+    credentialsProvider.pointee.getCredentials(credentialCallbackData: credentialCallbackData.pointee)
```

Comment:
So this is going to pass a raw pointer on to your customer code, would you rather just give them a callback to call with their results, and you can convert it back into gross C stuff? Otherwise, they have to work out the calling mechanism themselves, whereas calling a swift function is idiomatic.

Reply:
not sure what you mean. if you look at this PR: https://github.com/awslabs/aws-sdk-swift/pull/362 its how I expose it to the customer with just a simple protocol and I then I have a wrapper but it just calls the crt bindings nothing gross

Reply:
also im not passing the pointer but the pointee which is the swift type :)

Reply:
Ohhhhhhh